src/actions/commands/actions.ts

```diff
@@ -1840,10 +1861,123 @@ class CommandShowCommandLine extends BaseCommand {
   }
 }
 
+@RegisterAction
+class CommandNavigateInCommandlineOrSearchMode extends BaseCommand {
+  modes = [ModeName.CommandlineInProgress, ModeName.SearchInProgressMode];
+  keys = [['<left>'], ['<right>']];
+  runsOnceForEveryCursor() {
+    return this.keysPressed[0] === '\n';
+  }
+
+  public async exec(position: Position, vimState: VimState): Promise<VimState> {
+    const key = this.keysPressed[0];
+    const searchState = vimState.globalState.searchState!;
+    const statusbartext =
+      searchState !== undefined && searchState!.searchString
+        ? searchState!.searchString
+        : vimState.currentCommandlineText;
+    if (key === '<right>') {
+      vimState.statusBarCursorCharacterPos = Math.min(
+        vimState.statusBarCursorCharacterPos + 1,
+        statusbartext.length
+      );
+    } else if (key === '<left>') {
+      vimState.statusBarCursorCharacterPos = Math.max(vimState.statusBarCursorCharacterPos - 1, 0);
+    }
+    return vimState;
+  }
+}
+@RegisterAction
+class CommandTabInCommandline extends BaseCommand {
+  modes = [ModeName.CommandlineInProgress];
+  keys = ['<tab>'];
+  runsOnceForEveryCursor() {
+    return this.keysPressed[0] === '\n';
+  }
+
+  private autoComplete(completionItems: any, vimState: VimState) {
+    if (
+      commandLine.lastKeyPressed === '<tab>' &&
+      commandLine.autoCompleteIndex < completionItems.length
+    ) {
+      commandLine.autoCompleteIndex += 1;
+    } else if (/ /g.test(vimState.currentCommandlineText)) {
+      const search = <RegExpExecArray>/(?:.* .*\/|.* )(.*)/g.exec(vimState.currentCommandlineText);
+      commandLine.autoCompleteText = search[1];
+      commandLine.autoCompleteIndex = 0;
+    } else if (commandLine.lastKeyPressed !== '<tab>') {
+      commandLine.autoCompleteText = vimState.currentCommandlineText;
+      commandLine.autoCompleteIndex = 0;
+    }
+
+    completionItems = completionItems.filter(completionItem =>
+      completionItem.startsWith(commandLine.autoCompleteText)
+    );
+    if (commandLine.autoCompleteIndex >= completionItems.length) {
+      commandLine.autoCompleteIndex = 0;
+    }
+    let result = completionItems[commandLine.autoCompleteIndex];
+    if (result === vimState.currentCommandlineText && completionItems.length > 1) {
```

Comment:
This can be simplified to:

    if (result === vimState.currentCommandlineText) {
      result = completionItems[++commandLine.autoCompleteIndex % completionItems.length];
    }